### ISSUE TITLE
aws_elasticsearch_domain Advanced options are not applied properly

### ISSUE THREAD

**Issue by robertfirek** [hashicorp/terraform#3980]
When tried to apply the following advanced option on **aws_elasticsearch_domain**

```
(...)
advanced_options {
    "rest.action.multi.allow_explicit_index" = true
}
(...)
```

it was not applied properly in amazon (**rest.action.multi.allow_explicit_index** option was set to **false**).

Changing this option doesn't change plan (I've changed true to false) and it is impossible to control options.

State file:

```
(...)
"advanced_options.#": "1",
"advanced_options.rest.action.multi.allow_explicit_index": "false", 
(...)
```

Terraform version:
v0.6.5

**Comment by radeksimko** [comment]
I've got a hunch that this may be caused by the way we process the map for advanced_options:
https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/structure.go#L639

I'll try adding some extra logging and reproduce it.
Thanks for the report!

**Comment by radeksimko** [comment]
@robertfirek You can try specifying options as strings in the meantime:

```
advanced_options {
    "rest.action.multi.allow_explicit_index" = "true"
}
```

**Comment by robertfirek** [comment]
@radeksimko I've tried this approach before I wrote this issue. It doesn't work.

**Comment by gmorpheme** [comment]
I'm having the same issue - have tried with both booleans and strings with no luck.

**Comment by radeksimko** [comment]
After trying to reproduce this & running a few tests, I believe this is a core issue, described here https://github.com/hashicorp/terraform/issues/2143

Tags were something that a user could choose/change, but we cannot choose the map keys here, so the bug is affecting this resource more significantly. `advanced_options` are completely ignored at the moment due to this bug.

**Comment by genofedanzo** [comment]
Ran into this today.  Would love a fix.

**Comment by genofedanzo** [comment]
Also wanted to mention that from I testing I noticed that if you try to set advanced_options they end up being set to false  For instance, rest.action.multi.allow_explicit_index normally defaults to true but if you happen to manually set this to true it ends up being set as false.

**Comment by chiefy** [comment]
@genofedanzo also I believe the advanced options for ES domains are required, so it will essentially always be set to false :cry:

**Comment by genofedanzo** [comment]
@chiefy My work around was to not include advanced_options in terraform.  This made rest.action.multi.allow_explicit_index equal true (normally the default) and it looked like I could modify advanced_options in the AWS console without causing idempotent issues with the resource in terraform.  Not ideal but for now it won't prevent me from deploying ES to all of our environments.

**Comment by wnkz** [comment]
Also have this issue (`Terraform v0.6.9`)

**Comment by brentley** [comment]
+1 for trackinng purposes.

**Comment by radeksimko** [comment]
The work around until this gets fixed may include `local-exec` provisioner attached to this resource executing some AWS CLI commands to manage the advanced options.

It's probably not as portable as terraform code, but it does the job.

**Comment by chiefy** [comment]
@genofedanzo 's fix worked for me (just not including `advanced_options` in the template)

**Comment by moesy** [comment]
Confirmed as of today in v0.6.16 this bug still exists.

**Comment by radeksimko** [comment]
I just checked 0.7.0 and it may not work _exactly_ per the example in documentation, but dots in map keys are not the cause anymore.

By default we cast map values to strings before sending those to the API and `(boolean) true` is casted to `(string) "1"` which AWS API in turn translates into `false`.

The following example _does_ set `rest.action.multi.allow_explicit_index` to `true` (i.e. works as expected):

```hcl
resource "aws_elasticsearch_domain" "es" {
    domain_name = "tf-test"
    advanced_options {
        "rest.action.multi.allow_explicit_index" = "true"
        "indices.fielddata.cache.size" = 40
    }
}
```

![screen shot 2016-08-04 at 16 34 41](https://cloud.githubusercontent.com/assets/287584/17408008/6d3468f2-5a61-11e6-8688-17e7056f7436.png)


I'll relabel this to resource bug as we can still fix the map conversion in the context of this resource (i.e. probably turn `"1"` to `"true"`), but there's a simple workaround for this bug now in `0.7`. The other option would be to ask Amazon whether they can just take `"1"` as `true` 😃

**Comment by donnoman** [comment]
still a problem in 
```
› terraform --version
Terraform v0.8.5
```

**Comment by furhouse** [comment]
I've just ran into this issue with v0.8.5 as well, the following didn't work for me:

```
advanced_options {
    "rest.action.multi.allow_explicit_index" = "true"
}
```

**Comment by pracucci** [comment]
The following **works for us on terraform v. 0.8.8**:

```
advanced_options {
    "rest.action.multi.allow_explicit_index" = "true"
}
```

**Comment by wjessop** [comment]
Can confirm that @pracucci's fix works on v0.8.6
